include/functional.hpp

```diff
@@ -38,13 +38,27 @@ namespace std {
           return tag_invoke((Tag&&) tag, (Args&&) args...);
         }
       };
-    }
+
+      template <class>
+        void __tag_category();
+
+      template <class Tag>
+        typename Tag::tag_category __tag_category() requires true;
+    } // namespace __impl
+
     inline constexpr struct tag_invoke_t : __impl::tag_invoke_t {} tag_invoke {};
   }
 
   template<auto& Tag>
   using tag_t = decay_t<decltype(Tag)>;
 
+  template<class Tag>
+  using tag_category_t = decltype(__impl::__tag_category<Tag>());
+
+  template<class Tag, class... Categories>
+  concept tag_category_one_of =
+    (derived_from<tag_category_t<Tag>, Categories> ||...);
```

Comment:
I really do not like this

We should have a `receiver_querry` and a `sender_querry` concept.

They should simply inspect whether the alias is there and if so whether it is the right one. Do you ha e examples where we would need this dance?

Reply:
Edit, it seems one needs to pull the `same_as` out but yeah...
```cpp
template<class CPO>
concept receiver_query = requires {
    typename CPO::tag_category;
} && std::same_as<typename CPO::tag_category, query_tag>;
```

https://godbolt.org/z/raar7nsxG

Reply:
I like this suggestion better; it makes constraining the forwarding tag_invoke overloads easier than what you have in the PR right now.

Reply:
Surprisingly just `same_as` seems to also work even of there is no `tag_category`
https://godbolt.org/z/s3f4xnzoh